### PR TITLE
[WIP, Don't merge yet] Lsp module inside headless.

### PR DESCRIPTION
**Pull request by ndgnuh**
I'm having compilation error, can you help @hqnghi88 ?
```shell
[WIP, Build error] Add IdeSetup to Headless
Install GAMA project
Building  ummisco.gama.annotations
ok
return code 0
/home/hung/Projects/gama
Building  msi.gama.processor
ok
return code 0
/home/hung/Projects/gama
Building  msi.gama.parent
[ERROR] Internal error: java.lang.RuntimeException: org.osgi.framework.BundleException: Bundle msi.gama.ext cannot be resolved
[ERROR] Resolution errors:
[ERROR]    Bundle msi.gama.ext - Missing Constraint: Bundle-RequiredExecutionEnvironment: JavaSE-11
[ERROR] -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: org.osgi.framework.BundleException: Bundle msi.gama.ext cannot be resolved
Resolution errors:                                                                                                                                                    
   Bundle msi.gama.ext - Missing Constraint: Bundle-RequiredExecutionEnvironment: JavaSE-11                                                                           
                                                                                                                                                                      
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:120)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:64)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.RuntimeException: org.osgi.framework.BundleException: Bundle msi.gama.ext cannot be resolved
Resolution errors:                                                                                                                                                    
   Bundle msi.gama.ext - Missing Constraint: Bundle-RequiredExecutionEnvironment: JavaSE-11                                                                           
                                                                                                                                                                      
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getResolverState (OsgiBundleProject.java:266)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.resolveClassPath (OsgiBundleProject.java:164)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.resolveProject (DefaultTychoResolver.java:148)
    at org.eclipse.tycho.core.maven
```

### CODE REVIEW

**Comment by hqnghi88** [comment]
image
remove this line, tycho version is 2.2.0 for jdk 15+, it is set above

**Comment by ndgnuh** [comment]
Ok, let's continue, I got a new error
```shell
[WIP, Build error] Add IdeSetup to Headless
Install GAMA project
Building  ummisco.gama.annotations

ok
return code 0
/home/hung/Projects/gama
Building  msi.gama.processor
ok
return code 0
/home/hung/Projects/gama
Building  msi.gama.parent
Downloading from annot-local-repo: file:///home/hung/Projects/gama/msi.gama.lang.gaml.ide/../msi.gama.processor/processor/plugins/log4j/log4j/1.2.16/log4j-1.2.16.pom
Downloading from central: https://repo.maven.apache.org/maven2/log4j/log4j/1.2.16/log4j-1.2.16.pom
Downloaded from central: https://repo.maven.apache.org/maven2/log4j/log4j/1.2.16/log4j-1.2.16.pom (20 kB at 41 kB/s)
[ERROR] Internal error: org.apache.maven.artifact.InvalidArtifactRTException: For artifact {null:null:null:jar}: The groupId cannot be empty. -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: org.apache.maven.artifact.InvalidArtifactRTException: For artifact {null:null:null:jar}: The groupId cannot be empty.                                                                                                                                                              
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:120)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:64)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.artifact.InvalidArtifactRTException: For artifact {null:null:null:jar}: The groupId cannot be empty.
    at org.apache.maven.artifact.DefaultArtifact.validateIdentity (DefaultArtifact.java:130)
    at org.apache.maven.artifact.DefaultArtifact.<init> (DefaultArtifact.java:123)
    at org.apache.maven.bridge.MavenRepositorySystem.createArtifactX (MavenRepositorySystem.java:562)
    at org.apache.maven.bridge.MavenRepositorySystem.createArtifactX (MavenRepositorySystem.java:521)
    at org.apache.maven.bridge.MavenRepositorySystem.createArtifactX (MavenRepositorySystem.java:515)
    at org.apache.maven.bridge.MavenRepositorySystem.createArtifactX (MavenRepositorySystem.java:473)
    at org.apache.maven.bridge.MavenRepositorySystem.createArtifact (MavenRepositorySystem.java:80)
    at org.apache.maven.project.DefaultProjectBuilder.initProject (DefaultProjectBuilder.java:683)
    at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:188)
    at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:342)
    at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:299)
    at org.apache.maven.project.artifact.MavenMetadataSource.retrieveRelocatedProject (MavenMetadataSource.java:591)
    at org.apache.maven.project.artifact.MavenMetadataSource.retrieve (MavenMetadataSource.java:211)
    at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.recurse (DefaultLegacyArtifactCollector.java:551)
    at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.recurse (DefaultLegacyArtifactCollector.java:605)
    at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.recurse (DefaultLegacyArtifactCollector.java:605)
    at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.recurse (DefaultLegacyArtifactCollector.java:605)
    at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.recurse (DefaultLegacyArtifactCollector.java:605)
    at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.collect (DefaultLegacyArtifactCollector.java:148)
    at org.apache.maven.artifact.resolver.DefaultArtifactResolver.resolve (DefaultArtifactResolver.java:504)
    at org.apache.maven.repository.legacy.LegacyRepositorySystem.resolve (LegacyRepositorySystem.java:367)
    at org.apache.maven.DefaultProjectDependenciesResolver.resolveImpl (DefaultProjectDependenciesResolver.java:165)
    at org.apache.maven.DefaultProjectDependenciesResolver.resolve (DefaultProjectDependenciesResolver.java:81)
    at org.apache.maven.DefaultProjectDependenciesResolver.resolve (DefaultProjectDependenciesResolver.java:65)
    at org.eclipse.tycho.p2.resolver.P2DependencyResolver.collectPomDependencies (P2DependencyResolver.java:259)
    at org.eclipse.tycho.p2.resolver.P2DependencyResolver.computePreliminaryTargetPlatform (P2DependencyResolver.java:201)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.resolveProject (DefaultTychoResolver.java:120)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.lambda$resolveProjects$0 (TychoMavenLifecycleParticipant.java:127)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.accept (ForEachOps.java:183)
    at java.util.stream.WhileOps$1$1.accept (WhileOps.java:99)
    at java.util.ArrayList$ArrayListSpliterator.tryAdvance (ArrayList.java:1602)
    at java.util.stream.ReferencePipeline.forEachWithCancel (ReferencePipeline.java:127)
    at java.util.stream.AbstractPipeline.copyIntoWithCancel (AbstractPipeline.java:502)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:488)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:474)
    at java.util.stream.ForEachOps$ForEachOp.evaluateSequential (ForEachOps.java:150)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential (ForEachOps.java:173)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.forEach (ReferencePipeline.java:497)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.resolveProjects (TychoMavenLifecycleParticipant.java:158)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.afterProjectsRead (TychoMavenLifecycleParticipant.java:102)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:264)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:64)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/InternalErrorException
Something went wrong.
```

I don't what went wrong because all the artifactId are there?? Output of `mvn --version`
```shell
Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)
Maven home: /home/hung/Application/apache-maven
Java version: 15, vendor: Oracle Corporation, runtime: /usr/lib/jvm/openjdk15-bin
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.13.8_1", arch: "amd64", family: "unix"
```

File: msi.gama.headless/src/msi/gama/headless/runtime/Application.java [hunk 1]
```diff
@@ -41,8 +41,11 @@
 import org.osgi.framework.Bundle;
```
Comment (ndgnuh) [inline msi.gama.headless/src/msi/gama/headless/runtime/Application.java]: This commit name is misleading. It's actually resolving the conflict from new headless. Github handle this by merging the upstream to the PR.

**Comment by RoiArthurB** [comment]
@ndgnuh The developement of GAMA will soon be axed on GAMA 2 which have the main point to change the whole architecture.

I think, waiting for @AlexisDrogoul point of view, that it may be worth (and easier) to do that same work on for the branch `GAMA_2.0`. It would be usable for models developed with GAMA 1.8.2 (same langage), but it will be way easier for you as everything won't be spread all over the software as it is now on the 1.8.2 branch :)

**Comment by ndgnuh** [comment]
XText is complaining about un-registered language
```
SEVERE: Internal error: java.lang.reflect.InvocationTargetException
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
	at org.eclipse.lsp4j.jsonrpc.validation.ReflectiveMessageValidator.consume(ReflectiveMessageValidator.java:68)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 12 more
Caused by: java.lang.IllegalStateException: No Xtext languages have been registered. Please make sure you have added the languages's setup class in '/META-INF/services/org.eclipse.xtext.ISetup'
	at org.eclipse.xtext.ide.server.LanguageServerImpl.initialize(LanguageServerImpl.java:210)
	... 17 more
```

I have put `/META-INF/services/org.eclipse.xtext.ISetup` in several places and still nothing.

**Comment by AlexisDrogoul** [comment]
What is the content of the ISetup service you have put in `/META-INF/services/org.eclipse.xtext.ISetup` ?

**Comment by ndgnuh** [comment]
> What is the content of the ISetup service you have put in `/META-INF/services/org.eclipse.xtext.ISetup` ?

AFAIK that file contains the name of the Setup class, in this case, `GamlIdeSetup`. The newly generated project and the example projects found online are the same.

**Pull request closed without merging**